packages/grid/_modules_/grid/components/columnSelection/GridCellCheckboxRenderer.tsx

```diff
@@ -33,12 +33,18 @@ const GridCellCheckboxForwardRef = React.forwardRef<HTMLInputElement, GridCellPa
 
     const handleRef = useForkRef(checkboxElement, ref);
     const element = apiRef.current.getCellElement(id, field);
+    const shiftRef = React.useRef(false);
 
     const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-      apiRef.current.selectRow(id, event.target.checked, false);
+      if (shiftRef.current) {
+        apiRef.current.expandRowRangeSelection(id);
+      } else {
+        apiRef.current.selectRow(id, event.target.checked);
+      }
     };
 
     const handleClick = (event: React.MouseEvent<HTMLInputElement>) => {
+      shiftRef.current = event.shiftKey;
```

Comment:
The `onChange` does not have access to `shiftKey`
The `onClick` seems to always be called before `onChange` but if you have an alternative I'm interested. I did not wanted to listen to `keyDown` on every checkbox of the Grid.

Reply:
My alternative is to listen for the cell click event in the hook. https://github.com/mui-org/material-ui-x/pull/2456#discussion_r718031822